python/lsst/daf/butler/core/_butlerUri/s3.py

```diff
@@ -226,3 +229,75 @@ def transfer_from(self, src: ButlerURI, transfer: str = "copy",
         if transfer == "move":
             # Transactions do not work here
             src.remove()
+
+    @backoff.on_exception(backoff.expo, all_retryable_errors, max_time=max_retry_time)
+    def walk(self, file_filter: Optional[Union[str, re.Pattern]] = None) -> Iterator[Union[List,
+                                                                                           Tuple[ButlerURI,
+                                                                                                 List[str],
+                                                                                                 List[str]]]]:
+        """For dir-like URI, walk the directory returning matching files and
+        directories.
+
+        Parameters
+        ----------
+        file_filter : `str` or `re.Pattern`, optional
+            Regex to filter out files from the list before it is returned.
+
+        Yields
+        ------
+        dirpath : `ButlerURI`
+            Current directory being examined.
+        dirnames : `list` of `str`
+            Names of subdirectories within dirpath.
+        filenames : `list` of `str`
+            Names of all the files within dirpath.
+        """
+        # We pretend that S3 uses directories and files and not simply keys
+        if not self.isdir():
+            raise ValueError(f"Can not walk a non-directory URI: {self}")
+
+        if isinstance(file_filter, str):
+            file_filter = re.compile(file_filter)
+
+        s3_paginator = self.client.get_paginator('list_objects_v2')
+
+        # Limit each query to a single "directory" to match os.walk
+        # We could download all keys at once with no delimiter and work
```

Comment:
Because it's paginated, if you're going to scan them all eventually anyway, I think you'd be better off doing the simple version — except that you then wouldn't have the same result ordering as `os.walk()`.